### PR TITLE
npm install (ci) to WRKDIR, create distfile

### PR DESCRIPTION
**Pull request by Zirias**
* solves BUILD_AS_NON_ROOT in poudriere
 * solves multiple runs of npm install
 * solves writing to ${HOME} during fetch
 * reproducible distfile created with npm ci and mtree trickery

### CODE REVIEW

**Comment by tagattie** [comment]
Hi, thanks for creating the pull request. A reproducible node_modules archive is a good news!

The change mostly looks good to me. However, one thing I'm wondering is whether a distfile without MASTER_SITE is acceptable as an official FreeBSD port.

According to section 5.4.2 of Porter's Handbook:

>  MASTER_SITES must not be blank. It must point to the actual site hosting the distribution files. It cannot point to web archives, or the FreeBSD distribution files cache sites. The only exception to this rule is ports that do not have any distribution files. For example, meta-ports do not have any distribution files, so MASTER_SITES does not need to be set.

So it may be better saving the `pre-fetch` procedure for a port maintainer to create a node_modules archive in advance, and simply putting MASTER_SITE in the Makefile for pointing the prepared archive location.

Your thoughts?

**Comment by Zirias** [comment]
Well, you *have* `MASTER_SITES` here, but of course using a group for a distfile that's not defined in `MASTER_SITES` *is* a dirty hack and I' not sure whether FreeBSD would accept that. It's IMHO still better than using `npm install` to `${TMPDIR}` ;)

Preparing an own distribution file is simpler and solves the problem, but then you have to host it. IMHO it's better if a port uses only assets that are already available ... I see you already host a distfile for chromium though. I'm a bit unsure about all this :)

**Comment by tagattie** [comment]
Bad news...

I checked out `50-0x-npm-to-distfile` branch and tried it for myself. The size and checksum of generated tar.xz were different from those listed in the `distinfo`.

In my environment:

```
SHA256 (electron-npm-modules-5.0.1.tar.xz) = be605d5a30b22e5922e52a28800038e77758ce155e87025f57f462b45c22f483
SIZE (electron-npm-modules-5.0.1.tar.xz) = 15852088
```

:frowning_face:

**Comment by Zirias** [comment]
But you're getting the same (wrong) checksum on both? Maybe I made some stupid copy-and-paste error, I introduced the `NPM_TIMESTAMP` variable **after** testing myself. I can test again soon (tomorrow).

**Comment by tagattie** [comment]
Ah, OK. I'll wait for your result. In the meantime, I would like to propose some small changes.

File: Makefile [hunk 1]
```diff
@@ -8,7 +8,8 @@ MASTER_SITES=	https://github.com/tagattie/FreeBSD-Electron/releases/download/v5.
 		https://commondatastorage.googleapis.com/chromium-nodejs/:chromium_node \
 		https://commondatastorage.googleapis.com/chromium-fonts/:chromium_testfonts
 DISTFILES=	chromium-${CHROMIUM_VER}${EXTRACT_SUFX}:chromium \
-		${CHROMIUM_NODE_MODULES_HASH}:chromium_node
+		${CHROMIUM_NODE_MODULES_HASH}:chromium_node \
+		electron-npm-modules-${ELECTRON_VER}.tar.xz:prefetch
```
Comment (tagattie) [inline Makefile]:
```suggestion
		electron-npm-modules-${ELECTRON_VER}${EXTRACT_SUFX}:prefetch
```

File: Makefile [hunk 2]
```diff
@@ -149,11 +150,24 @@ PULSEAUDIO_VARS_OFF=	GN_ARGS+=use_pulseaudio=false
...
 
-post-fetch:
-	${RM} -r ${TMPDIR}/npm-cache
-	${MKDIR} ${TMPDIR}/npm-cache
-	${CP} ${FILESDIR}/package.json ${FILESDIR}/package-lock.json ${TMPDIR}/npm-cache
-	cd ${TMPDIR}/npm-cache && npm install --verbose --no-progress
+NPM_TIMESTAMP=	1557238990
+
+pre-fetch:
+	if [ ! -f ${DISTDIR}/electron-npm-modules-${ELECTRON_VER}.tar.xz ]; \
```
Comment (tagattie) [inline Makefile]:
```suggestion
	if [ ! -f ${DISTDIR}/electron-npm-modules-${ELECTRON_VER}${EXTRACT_SUFX} ]; \
```

File: Makefile [hunk 3]
```diff
@@ -149,11 +150,24 @@ PULSEAUDIO_VARS_OFF=	GN_ARGS+=use_pulseaudio=false
...
+		then ${MKDIR} ${WRKDIR}/npm-cache; \
+		${CP} ${FILESDIR}/package.json \
+			${FILESDIR}/package-lock.json ${WRKDIR}/npm-cache; \
+		cd ${WRKDIR}/npm-cache && \
+		HOME=${WRKDIR} npm ci --verbose --no-progress && \
```
Comment (tagattie) [inline Makefile]:
```suggestion
		${SETENV} HOME=${WRKDIR} npm ci --verbose --no-progress && \
```

File: Makefile [hunk 4]
```diff
@@ -149,11 +150,24 @@ PULSEAUDIO_VARS_OFF=	GN_ARGS+=use_pulseaudio=false
...
+		${MV} node_modules npm_modules; \
+		mtree -cbnSp npm_modules | mtree -C | sed \
```
Comment (tagattie) [inline Makefile]:
```suggestion
		${MTREE_CMD} -cbnSp npm_modules | ${MTREE_CMD} -C | ${SED} \
```

File: Makefile [hunk 5]
```diff
@@ -149,11 +150,24 @@ PULSEAUDIO_VARS_OFF=	GN_ARGS+=use_pulseaudio=false
...
+			-e 's:time=[0-9.]*:time=${NPM_TIMESTAMP}.000000000:' \
+			-e 's:\([gu]id\)=[0-9]*:\1=0:g' \
+			-e 's:^\.:./npm_modules:' > npm_modules.mtree; \
+		tar cJf ${DISTDIR}/electron-npm-modules-${ELECTRON_VER}.tar.xz \
```
Comment (tagattie) [inline Makefile]:
```suggestion
		${TAR} cJf ${DISTDIR}/electron-npm-modules-${ELECTRON_VER}${EXTRACT_SUFX} \
```

File: Makefile (hunk 3 above)
Comment (Zirias) [reply Makefile]: I'm not sure this does what's expected? In a bourne-compatible shell, an environment variable is set for just the following command when you write it in front. Does using `setenv` do the same thing, and if so, what's the benefit?

**Comment by Zirias** [comment]
Ok I just tested again, doing `makesums` on the result. It's again

    SHA256 (electron-npm-modules-5.0.1.tar.xz) = d687aed401d32d6df7b10ecafa15fe4391a58a88454b5718ec026ac609c1580e
    SIZE (electron-npm-modules-5.0.1.tar.xz) = 15851384

So, the good news is: this method of creating a tarball **is** reproducible. The bad news is, there's still a difference between our machines. I'm running `12.0-RELEASE-p3` on amd64 as well, so it isn't related to the host system :(

File: Makefile (hunk 3 above)
Comment (tagattie) [reply Makefile]: The value of `SETENV` is `/usr/bin/env` and it does the same thing according to the man page.

I have followed a (seemingly) FreeBSD ports convention although it seems unnecessary.
(`${SETENV} environment-variables command` is seen here and there in bsd.*.mk and other ports' Makefiles).

This port has `${SETENV}` in `pre-configure` and `do-configure` targets as well.

Comment (Zirias) [reply Makefile]: I see, so this is for (bourne) shells that don't support it themselves? Didn't think/know something like that existed. But then, it's pointless until I can find out why you're getting a different tarball....

**Comment by tagattie** [comment]
File flags may have something to do with the difference.

``` shell-session
$ head -n 3 npm_modules.mtree
./npm_modules type=dir uid=0 gid=0 mode=0755 nlink=925 time=1557238990.000000000 flags=uarch
./npm_modules/.bin type=dir uid=0 gid=0 mode=0755 nlink=2 time=1557238990.000000000 flags=uarch
./npm_modules/.bin/JSONStream type=link uid=0 gid=0 mode=0755 nlink=1 link=../JSONStream/bin.js time=1557238990.000000000 flags=uarch
```

The directory is on a ZFS filesystem and all the mtree entries have `flags=uarch`. I tried `mtree` command with other machine's some directory on a UFS filesystem and the entries have `flags=none`.

What does it look like on your side?

**Comment by Zirias** [comment]
I *guess* you found it, great, I completely forgot `tar` also archives flags. I'm using ZFS as well, still I had "flags=none" everywhere. Please test with the updated `Makefile` that *forces* flags to be "none", thanks :)

**Comment by tagattie** [comment]
Finally I got the same result as yours:

```
SHA256 (electron-npm-modules-5.0.1.tar.xz) = d687aed401d32d6df7b10ecafa15fe4391a58a88454b5718ec026ac609c1580e
SIZE (electron-npm-modules-5.0.1.tar.xz) = 15851384
```

Yay :smile: I will merge the changes shortly. I do appreciate your time and work on this!

**Pull request merged**